### PR TITLE
Add automated gallery content maintenance

### PR DESCRIPTION
**Pull request by jaydestro**
## Summary

Adds production-ready, human-gated gallery maintenance for `AzureCosmosDB/gallery` while keeping the site fully static:

- audits all live catalog entries and discovers bounded candidates from approved blogs, YouTube channels, GitHub organizations, and Microsoft Learn
- uses a tool-free GitHub Copilot CLI classification step with strict schema and authoritative-index validation
- requires the Azure Cosmos DB Agent Kit `cosmosdb-best-practices` skill and the repository `humanizer` skill
- runs automatically once a week, Monday at 06:17 UTC; manual dispatch remains available for setup and recovery
- rebuilds one draft maintenance proposal from current `main`; automation never merges or pushes to `main`
- supports deterministic `Reject: A1, U1, R1` maintainer comments and revalidates each revised proposal
- requests review from `jagord_microsoft` only when a new maintenance PR is created
- excludes Azure DocumentDB content and records detailed proof for every proposed retirement
- publishes catalog cards newest-first without changing `featured` metadata
- rejects localized Microsoft documentation URLs such as `/en-us/` in live and retired catalog records
- migrates dependency and Pages workflows from Yarn/PAT-based `gh-pages` publication to npm and official GitHub Pages artifact deployment

## Documentation

- adds the architecture and implementation proposal
- adds the source eligibility, product scope, canonical URL, and retirement policy
- adds the operator runbook covering token creation, secret installation, permissions, skill setup, notification routing, first-run enablement, review commands, failure recovery, rollback, and Copilot CLI updates
- updates README and contributing guidance to describe the upstream URL, npm validation, and human review boundary

### 1. Create `COPILOT_GITHUB_TOKEN`

Use a maintained automation identity with an active GitHub Copilot subscription and permission to use Copilot CLI.

1. Open **Settings > Developer settings > Personal access tokens > Fine-grained tokens > Generate new token**.
2. Set **Resource owner** to the automation identity's personal account, not `AzureCosmosDB`.
3. Set **Repository access** to **Public repositories**.
4. Under **Account permissions**, add **Copilot Requests** only.
5. Set a policy-compliant expiration, generate the token, and copy the value immediately.

### 2. Create `GALLERY_UPDATE_TOKEN`

Use a maintained automation identity that is a member of `AzureCosmosDB` and can create branches and pull requests in this repository.

1. Create another fine-grained token.
2. Set **Resource owner** to **AzureCosmosDB**.
3. Set **Repository access** to **Only select repositories > gallery**.
4. Grant only **Contents: Read and write** and **Pull requests: Read and write**.
5. Set a policy-compliant expiration, generate the token, and copy the value immediately.
6. If the token is `pending`, an organization owner must approve it before the workflow can publish.

These must be separate tokens: Copilot Requests is a personal-account permission, while repository write access is organization-scoped.

### 3. Install the Actions secrets

In **AzureCosmosDB/gallery > Settings > Secrets and variables > Actions > Secrets**, add both values with these exact names:

- `COPILOT_GITHUB_TOKEN`
- `GALLERY_UPDATE_TOKEN`

Alternatively, use commands that prompt for each value without placing it in shell history:

```shell
gh secret set COPILOT_GITHUB_TOKEN --repo AzureCosmosDB/gallery
gh secret set GALLERY_UPDATE_TOKEN --repo AzureCosmosDB/gallery
gh secret list --repo AzureCosmosDB/gallery
```

Never paste token values into issues, pull requests, workflow inputs, command arguments, or logs.

### 4. Configure notification delivery

The workflow requests `jagord_microsoft` as reviewer only when it creates a new maintenance PR. In that account:

1. Verify `jagord@microsoft.com` under **Settings > Emails**.
2. Enable email delivery for participating and review-request notifications under **Settings > Notifications**.
3. Route `AzureCosmosDB` organization notifications to `jagord@microsoft.com`.
4. Confirm the first generated PR lists `jagord_microsoft` under **Reviewers** and that the email arrives.

No SMTP or mail-service secret is required.

### 5. Enable and verify workflows

1. Approve the pinned GitHub Actions used by the repository.
2. In **Settings > Pages**, change **Build and deployment > Source** from `gh-pages` to **GitHub Actions**.
3. Manually run **Audit gallery content** once. Confirm Copilot classification reports `complete`, both required skills pass preflight, and an eligible proposal can create or update the draft PR.
4. Review the first artifact and draft PR before relying on the weekly schedule.
5. Confirm a normal merge to `main` deploys `https://azurecosmosdb.github.io/gallery/` through **Deploy to GitHub Pages**.

## Validation

- `npm ci`
- `npm run test:gallery-audit` (62 passed)
- `npm run gallery:audit:fixtures`
- `npm run build`
- pinned Agent Kit download checksum and real Copilot CLI discovery of both required skills
- all workflow YAML parsed successfully
- exactly one automatic cron schedule
- no `jaydestro` or `jaydestro.github.io` references in automation, scripts, or documentation
- no localized Microsoft documentation URLs in live or retired catalogs
- no changes to `static/templates.json`; `static/retired-templates.json` is introduced as an empty infrastructure ledger
- no workflow can merge, enable automerge, or push directly to `main`

### CODE REVIEW

**Comment by jaydestro** [comment]
@sajeetharan The Copilot review findings have been addressed and all threads are resolved. The current head passes 62 tests, fixture validation, the production build, YAML lint, and upstream CI. Please review, approve, and merge when ready. I will handle the required token setup after merge and reach out if I need organization-owner assistance.

File: scripts/gallery-audit/review-command.mjs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  const baseOrder = new Map(baseCatalog.map((entry, index) => [entry.title, index]));
+  const additions = live.filter((entry) => !baseOrder.has(entry.title));
+  const existing = live
+    .filter((entry) => baseOrder.has(entry.title))
+    .sort((left, right) => baseOrder.get(left.title) - baseOrder.get(right.title));
+  return { liveCatalog: sortCatalogForPublishing([...additions, ...existing]), retiredCatalog: retired };
+}
+
+export function summarizeCatalogDiff({ baseCatalog, catalog, baseRetiredCatalog, retiredCatalog, generatedAt }) {
+  const baseByTitle = new Map(baseCatalog.map((entry) => [entry.title, entry]));
```
Comment (sajeetharan) [inline scripts/gallery-audit/review-command.mjs]: Can we avoid using the title as the unique key here? The catalog already has entries with the same title but different URLs, so this can match the wrong item and report a false URL update. Please use a unique identifier and add a test for duplicate titles.

File: scripts/gallery-audit/core.mjs [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+      const metadata = JSON.parse(apiResult.body);
+      if (metadata.private) return { ...apiResult, outcome: 'review', reason: 'github-private' };
+      if (metadata.disabled) return { ...apiResult, outcome: 'review', reason: 'github-disabled' };
+      if (metadata.archived) return { ...apiResult, outcome: 'review', reason: 'github-archived' };
+      if (repository.hasPath) {
+        const { headers: _headers, ...pathOptions } = options;
+        const pathResult = await boundedGet(value, policy, { ...pathOptions, allowedHostnames: ['github.com'] });
+        return { ...pathResult, ...statusResult(pathResult) };
+      }
+      return { ...apiResult, finalUrl: metadata.html_url ?? value, outcome: 'healthy', reason: 'github-active' };
```
Comment (sajeetharan) [inline scripts/gallery-audit/core.mjs]: This always returns healthy, even when GitHub redirects a renamed or transferred repository. That means the old URL will never be updated in the catalog. Can we return redirected when the API redirects or html_url differs from the current URL, and add a test for it?

File: .github/workflows/apply-gallery-review.yml [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+concurrency:
+  group: gallery-content-audit
+  cancel-in-progress: false
+
+jobs:
+  apply:
+    if: >-
+      github.event.issue.pull_request &&
+      contains(github.event.comment.body, 'Reject:') &&
+      contains(fromJSON('["OWNER","MEMBER","COLLABORATOR"]'), github.event.comment.author_association)
```
Comment (sajeetharan) [inline .github/workflows/apply-gallery-review.yml]: MEMBER and COLLABORATOR do not always mean the user has write access. A read-only member could trigger this workflow and use the privileged token to update the branch. Can we check the commenter’s repository permission and allow only write, maintain, or admin access?

File: .github/workflows/audit-gallery-content.yml [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+            fi
+            if git ls-remote --exit-code --heads origin "$UPDATE_BRANCH" > /dev/null; then
+              git push --delete origin "$UPDATE_BRANCH"
+            fi
+            exit 0
+          fi
+
+          git add static/templates.json static/retired-templates.json
+          git commit -m "chore: refresh gallery content"
+          git push --force-with-lease origin "HEAD:$UPDATE_BRANCH"
```
Comment (sajeetharan) [inline .github/workflows/audit-gallery-content.yml]: This force-push rebuilds the branch from main, so it can remove the commits that recorded a reviewer’s rejected items. Those items may then appear again in the next audit. Can we preserve the rejection decisions, or create a new draft PR for each run?

File: .github/workflows/audit-gallery-content.yml [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+          pr_number=$(gh pr list --head "$UPDATE_BRANCH" --base main --state open --json number --jq '.[0].number')
+          if [ -n "$pr_number" ]; then
+            gh pr edit "$pr_number" --title "chore: refresh gallery content" --body-file output/gallery-content-review/promotion-summary.md
+          else
+            gh pr create --draft --base main --head "$UPDATE_BRANCH" --reviewer "$NOTIFY_REVIEWER" --title "chore: refresh gallery content" --body-file output/gallery-content-review/promotion-summary.md
```
Comment (sajeetharan) [inline .github/workflows/audit-gallery-content.yml]: Requesting a reviewer does not require their approval before merge. Since main is currently unprotected, this does not enforce the human approval gate described in the PR. Can we add the required branch protection and stale-approval settings before enabling this workflow?

**Review by sajeetharan — CHANGES_REQUESTED** [review]
The overall direction looks good. However, I’m requesting changes because a few safeguards need to be addressed before this is enabled:

File: scripts/gallery-audit/review-command.mjs (hunk 1 above)
Comment (jaydestro) [reply scripts/gallery-audit/review-command.mjs]: Addressed in b6878dc: proposal actions now match entries by unique source URL, and diff regeneration pairs URL changes by unchanged catalog metadata. Added duplicate-title regression coverage.

File: scripts/gallery-audit/core.mjs (hunk 2 above)
Comment (jaydestro) [reply scripts/gallery-audit/core.mjs]: Addressed in b6878dc: GitHub API redirects or a changed canonical html_url now return redirected with the replacement URL. Added rename/transfer regression coverage.

File: .github/workflows/apply-gallery-review.yml (hunk 3 above)
Comment (jaydestro) [reply .github/workflows/apply-gallery-review.yml]: Addressed in b6878dc: the workflow now queries the commenter effective repository permission and permits only write, maintain, or admin before the privileged token is used.

File: .github/workflows/audit-gallery-content.yml (hunk 4 above)
Comment (jaydestro) [reply .github/workflows/audit-gallery-content.yml]: Addressed in b6878dc: every audit run now creates a new run-specific branch and draft PR; it no longer force-pushes over the branch containing review rejection commits.

File: .github/workflows/audit-gallery-content.yml (hunk 5 above)
Comment (jaydestro) [reply .github/workflows/audit-gallery-content.yml]: Addressed in b6878dc and repository settings: active ruleset Protect main (ID 23827086) requires one approval, dismisses stale approvals, requires approval after the latest push, and requires resolved threads. The audit also fails closed if those controls are absent.

File: scripts/gallery-audit/core.mjs [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+      return false;
+    }
+  });
+}
+
+export async function auditCatalog(catalog, policy, options = {}) {
+  validateCatalog(catalog);
+  const duplicates = findDuplicates(catalog, policy.trackingParameters);
+  const scannedAt = (options.now ?? new Date()).toISOString();
+  const checker = options.checker ?? ((url) => checkUrl(url, policy));
```
[the remark this answers is not in the document]
Comment (jaydestro) [reply scripts/gallery-audit/core.mjs]: Addressed in 01b73f3: cross-host redirect destinations are marked review-only during audit, and promotion independently enforces same-host destinations.

File: scripts/gallery-audit/promotion.mjs [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+  const knownUrls = new Set([...catalog, ...retiredCatalog].map((entry) => urlFingerprint(entry.source, policy.trackingParameters)));
+  const additions = [];
+  const skippedAdditions = [];
+  const updates = [];
+  const updatedCatalog = catalog.map((entry) => ({ ...entry }));
+
+  for (const auditEntry of auditReport.entries) {
+    if (auditEntry.outcome !== 'redirected' || !auditEntry.finalUrl) continue;
+    const original = updatedCatalog[auditEntry.catalogIndex];
+    if (!original || original.source !== auditEntry.url) continue;
```
[the remark this answers is not in the document]
Comment (jaydestro) [reply scripts/gallery-audit/promotion.mjs]: Addressed in 01b73f3: automatic URL updates now require a normalized same-host destination; unapproved redirects remain review-only.

File: scripts/gallery-audit/promotion.mjs [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    `  - Preview: ${entry.preview?.trim() || '**MISSING**'}`,
+    `  - Source: ${entry.source?.trim() || '**MISSING**'}`,
+  ];
+  return [
+    '# Automated gallery content update', '',
+    `Generated: ${generatedAt}`, '',
+    'Comment with item IDs to reject proposed changes, for example: `Reject: A1, U1, R1`.', '',
+    `Additions: ${result.additions.length}`, '',
+    ...result.additions.flatMap((entry, index) => [
+      `- **A${index + 1}** Add [${entry.title}](${entry.source})`,
```
[the remark this answers is not in the document]
Comment (jaydestro) [reply scripts/gallery-audit/promotion.mjs]: Addressed in 01b73f3: proposal identity now comes from base64url-encoded JSON markers, not rendered Markdown. Visible titles and URLs are safely rendered.

File: scripts/gallery-audit/review-command.mjs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  }
+  const retirements = retiredCatalog.filter((entry) => !baseRetiredSources.has(entry.source));
+  const removedWithoutRetirement = unmatchedBase.filter((entry) => !retirements.some((retired) => retired.source === entry.source));
+  if (removedWithoutRetirement.length > 0) throw new Error(`Catalog entries were removed without retirement records: ${removedWithoutRetirement.map((entry) => entry.title).join(', ')}`);
+  const cardDetails = (entry) => [
+    `  - Description: ${entry.description?.trim() || '**MISSING**'}`,
+    `  - Author: ${Array.isArray(entry.author) ? entry.author.join(', ') : (entry.author?.trim() || '**MISSING**')}`,
+    `  - Date: ${entry.date?.trim() || '**MISSING**'}`,
+    `  - Tags: ${entry.tags?.length ? entry.tags.join(', ') : '**MISSING**'}`,
+    `  - Website: ${entry.website?.trim() || '**MISSING**'}`,
```
[the remark this answers is not in the document]
Comment (jaydestro) [reply scripts/gallery-audit/review-command.mjs]: Addressed in 01b73f3: regenerated summaries use the same structured markers and shared escaping helpers; forged Markdown items are ignored by regression test.

File: .github/workflows/apply-gallery-review.yml [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    env:
+      PR_NUMBER: ${{ github.event.issue.number }}
+      REVIEW_COMMAND: ${{ github.event.comment.body }}
+    steps:
+      - name: Authorize commenter
+        env:
+          GH_TOKEN: ${{ github.token }}
+          COMMENTER: ${{ github.event.comment.user.login }}
+        run: |
+          permission=$(gh api "repos/$GITHUB_REPOSITORY/collaborators/$COMMENTER/permission" --jq .permission)
```
[the remark this answers is not in the document]
Comment (jaydestro) [reply .github/workflows/apply-gallery-review.yml]: Addressed in 01b73f3: authorization uses GALLERY_UPDATE_TOKEN, and the runbook now requires Administration read solely for effective collaborator permission checks.

File: .github/workflows/audit-gallery-content.yml [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+            fi
+          done
+          test "$protected" = "true" || {
+            echo "main must require an approving review, stale-review dismissal, approval after the latest push, and resolved review threads."
+            exit 1
+          }
+
+      - name: Prepare per-run automation branch
+        env:
+          UPDATE_BRANCH: automation/gallery-content-updates-${{ github.run_id }}
```
[the remark this answers is not in the document]
Comment (jaydestro) [reply .github/workflows/audit-gallery-content.yml]: Addressed in 01b73f3: automation branches now include both github.run_id and github.run_attempt, so reruns receive distinct branches and PRs.

File: scripts/gallery-audit/promotion.mjs [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    let previousFingerprint;
+    let finalFingerprint;
+    try {
+      previousFingerprint = urlFingerprint(original.source, policy.trackingParameters);
+      finalFingerprint = urlFingerprint(auditEntry.finalUrl, policy.trackingParameters);
+    } catch {
+      continue;
+    }
+    if (previousFingerprint === finalFingerprint) continue;
+    const conflicts = updatedCatalog.some((entry, index) => index !== auditEntry.catalogIndex && urlFingerprint(entry.source, policy.trackingParameters) === finalFingerprint);
```
[the remark this answers is not in the document]
Comment (jaydestro) [reply scripts/gallery-audit/promotion.mjs]: Addressed in 01b73f3: redirect conflicts are checked against knownUrls, which includes both live and retired catalog sources.

File: scripts/gallery-audit/promotion.mjs [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+    if (conflicts) continue;
+    const previousUrl = original.source;
+    original.source = auditEntry.finalUrl;
```
[the remark this answers is not in the document]
Comment (jaydestro) [reply scripts/gallery-audit/promotion.mjs]: Addressed in 01b73f3: redirect destinations are normalized before conflict checks, catalog assignment, and update reporting.

File: scripts/gallery-audit/promotion.mjs [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+    readJson('output/gallery-content-review/article-candidates.json'),
+    readJson('output/gallery-content-review/audit-report.json'),
+    readJson('output/gallery-content-review/run-metadata.json'),
+    readJson('.github/gallery-audit/sources.json'),
+    readJson('.github/gallery-audit/policy.json'),
+  ]);
+  if (!metadata.complete || metadata.copilot?.status !== 'complete') {
+    throw new Error('Promotion requires a complete deterministic scan and Copilot classification');
+  }
+  const result = planCatalogPromotion({ catalog, retiredCatalog, candidateReport, auditReport, sourcesDocument, policy, now });
```
[the remark this answers is not in the document]
Comment (jaydestro) [reply scripts/gallery-audit/promotion.mjs]: Addressed in 01b73f3: both promoted live and retired catalogs are validated before any write; localized replacement URLs now fail publication.

**Comment by jaydestro** [comment]
@sajeetharan Your five requested safeguards have been addressed in `b6878dc`, including unique URL-based identity, GitHub rename/transfer handling, effective commenter permission checks, per-run maintenance PR branches, and enforced `main` approval protections.

A follow-up GitHub Copilot code review was run against those changes. All nine additional findings from that review were addressed in `01b73f3`, with a specific response left on each thread. Local validation passes 71 tests, fixture discovery, and the production build.

This should now be ready to go. Please review the latest changes and approve for merge when satisfied.  I do not have write permissions to main.  I also will have to set up tokens in GitHub actions.

**Review by sajeetharan — APPROVED** [review]
